qiskit/backends/ibmq/ibmqbackend.py

```diff
@@ -53,7 +53,9 @@ def run(self, q_job):
         Returns:
             IBMQJob: an instance derived from BaseJob
         """
-        return IBMQJob(q_job, self._api, not self.configuration['simulator'])
+        ibmq_job = IBMQJob(q_job, self._api, not self.configuration['simulator'])
+        ibmq_job.submit()
```

Comment:
You made a point here that makes a lot of sense and would worth a review it in a future iteration: maybe the Executor instance might reside elsewhere, as a Job will never create more than one task at the same time, so there's no need to create a thread pool in there... but this will require a proper design, let's put it in our TOOD list for now.

I created the `submit` method because there seems to be two kind of situations where a job could be created: 
1. From a circuit being build by the user.
2. From information retrieved from the API. 
In the first one a submit procedure makes sense, but not in the second one. By submitting in the constructor, you are forced to treat the latter as an exceptional case, having to create a static constructor that repeats a lot of the initialization code and adds extra complexity. My initial plan was to unify Job constructor, because it's pretty weird for me that we have to instantiate jobs in a different way if the info is coming from the API or from the circuit, this is probably an indication that something is wrong here. Why the API is not giving us the same information we have at circuit creation time? I had a look into it, and the main problem is that we are passing a `circuit` in it's JSON representation (via QuantumJob.q_obj) in the first case, but in the second case we don't have this `circuit` representation, we have the `qasm` representation of the circuit. But, the job `_submit()` method transforms the JSON `circuit` representation into `qasm`!!. If we fix this, we will be able to have just one constructor for every situation, but as explained above, the `submit` procedure should be done explicitly because we have two different situations.

Reply:
This makes sense. As long as backend.run() returns an already submitted job, as I think it does in your PR, then the user doesn't have to call `submit` after calling `run`. Maybe put this into a PR on it's own?